### PR TITLE
[7.0] SEPA SWISS LOCALIZATION

### PR DESCRIPTION
**Pull request by abaghirli**

### CODE REVIEW

File: l10n_ch_sepa_credit_transfer/__init__.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+#    Author: Yannick Vaucher
+#    Copyright 2011 Camptocamp SA
```
Comment (yvaucher) [inline l10n_ch_sepa_credit_transfer/__init__.py]: You can change this to put you as author and 2017.

File: l10n_ch_sepa_credit_transfer/__openerp__.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+You'll found the wizard in payment order.
+
+This module designed for Swiss payment is also generic for euro payments
+as the Swiss standards are slightly different.
+It intends to be reusable to accept new definition of specific standards
+by country.
+
+It currently supports the "pain.001" norm for Credit Transfert Initiation.
+
+""",
```
Comment (yvaucher) [inline l10n_ch_sepa_credit_transfer/__openerp__.py]: This description needs to be moved in README.rst

File: l10n_ch_sepa_credit_transfer/__openerp__.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+#    Author: Yannick Vaucher
+#    Copyright 2011 Camptocamp SA
```
Comment (yvaucher) [inline l10n_ch_sepa_credit_transfer/__openerp__.py]: You can change author and year here

File: l10n_ch_sepa_credit_transfer/wizard/export_sepa.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+            gen_args = {
+                'bic_xml_tag': bic_xml_tag,
+                'name_maxsize': name_maxsize,
+                'convert_to_ascii': convert_to_ascii,
+                'payment_method': 'TRF',
+                'pain_flavor': pain_flavor,
+                'sepa_export': sepa_export,
+                'file_obj': self.pool['banking.export.sepa'],
+                'pain_xsd_file':
+                'account_banking_sepa_credit_transfer/data/%s.xsd'
```
Comment (yvaucher) [inline l10n_ch_sepa_credit_transfer/wizard/export_sepa.py]: Here you can do:

``` python
if pain_flavor ...:
   module = ...
else:
   module = ...
gen_args = {
            [...]
            'pain_xsd_file': '%s/data/%s.xsd' % (module, pain_flavor)
}
```

File: l10n_ch_sepa_direct_debit/wizard/export_sdd.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+#    SEPA Credit Transfer module for OpenERP
```
Comment (yvaucher) [inline l10n_ch_sepa_direct_debit/wizard/export_sdd.py]: s/Credit Transfer/Direct Debit

File: l10n_ch_sepa_direct_debit/wizard/export_sdd.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+from openerp.osv import orm, fields
+from openerp.tools.translate import _
+from lxml import etree
+
+
+class BankingExportSepaWizard(orm.TransientModel):
+    _inherit = 'banking.export.sdd.wizard'
+
+
+def create_sepa(self, cr, uid, ids, context=None):
```
Comment (yvaucher) [inline l10n_ch_sepa_direct_debit/wizard/export_sdd.py]: Indentation is wrong here

File: l10n_ch_sepa_direct_debit/wizard/export_sdd.py [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+            'bic_xml_tag': bic_xml_tag,
+            'name_maxsize': name_maxsize,
+            'convert_to_ascii': convert_to_ascii,
+            'payment_method': 'DD',
+            'pain_flavor': pain_flavor,
+            'sepa_export': sepa_export,
+            'file_obj': self.pool['banking.export.sdd'],
+            'pain_xsd_file':
+                'account_banking_sepa_direct_debit/data/%s.xsd' % pain_flavor,
+        }
```
Comment (yvaucher) [inline l10n_ch_sepa_direct_debit/wizard/export_sdd.py]: Same comment than for credit transfer. Here you could have a placeholder for module

**Review by yvaucher — CHANGES_REQUESTED** [review]
I only see XSD validation changes. Are some commit missing ? It seems that an equivalence to 
https://github.com/OCA/l10n-switzerland/tree/10.0/l10n_ch_pain_base
is missing.


Indent is wrong in l10n_ch_sepa_direct_debit/wizard/export_sdd.py

File: l10n_ch_sepa_credit_transfer/__openerp__.py (hunk 2 above)
Comment (abaghirli) [reply l10n_ch_sepa_credit_transfer/__openerp__.py]: in version 7 there is no Readme.rst files in modules.

File: l10n_ch_sepa_credit_transfer/data/payment_type_sepa_sct.xml [hunk 8]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+<data noupdate="0">
+
+
+<!-- TODO In the suitable_bank_types field, we should restrict to IBAN type and Rib+IBAN of l10n_fr_rib ? -->
```
Comment (yvaucher) [inline l10n_ch_sepa_credit_transfer/data/payment_type_sepa_sct.xml]: Not sure to get that comment, l10n_fr_rib is related to france localization not switzerland one.

File: l10n_ch_sepa_credit_transfer/wizard/export_sepa.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+            if pain_flavor in pain_03_to_05:
+                nb_of_transactions_2_4.text = str(transactions_count_2_4)
+                control_sum_2_5.text = '%.2f' % amount_control_sum_2_5
+
+        if pain_flavor in pain_03_to_05:
+            nb_of_transactions_1_6.text = str(transactions_count_1_6)
+            control_sum_1_7.text = '%.2f' % amount_control_sum_1_7
+        else:
+            nb_of_transactions_1_6.text = str(transactions_count_1_6)
+            control_sum_1_7.text = '%.2f' % amount_control_sum_1_7
```
Comment (yvaucher) [inline l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:426]: If and else are doing the exact same thing?

File: l10n_ch_sepa_direct_debit/__openerp__.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+{
+    "name": "Switzerland - SEPA Electronic Payment File",
+    "summary": "Generate pain.001 Credit Transfert Files for your payments",
```
Comment (yvaucher) [inline l10n_ch_sepa_direct_debit/__openerp__.py]: Name and summary to fix.

File: l10n_ch_sepa_direct_debit/__openerp__.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    "version": "1.0",
```
Comment (yvaucher) [inline l10n_ch_sepa_direct_debit/__openerp__.py]: version should be 7.0.1.0.0

File: l10n_ch_sepa_direct_debit/__openerp__.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+It intends to be reusable to accept new definition of specific standards
+by country.
+
+It currently supports the "pain.008" norm for Direct Debit Initiation.
+
+""",
+    "author": "Camptocamp,Odoo Community Association (OCA)",
+    "depends": [
+        "account",
+        "l10n_ch",
```
Comment (yvaucher) [inline l10n_ch_sepa_direct_debit/__openerp__.py]: We might not need to depend on l10n_ch as it is only the chart of account.

File: l10n_ch_sepa_direct_debit/data/payment_type_sepa_sct.xml [hunk 13]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+<data noupdate="0">
+
+
+<!-- TODO In the suitable_bank_types field, we should restrict to IBAN type and Rib+IBAN of l10n_fr_rib ? -->
```
Comment (yvaucher) [inline l10n_ch_sepa_direct_debit/data/payment_type_sepa_sct.xml]: Same comment as in other module.
This is related to question about france localisation.

**Review by yvaucher — CHANGES_REQUESTED** [review]
Please check travis for lint

Trying manually, I cannot generate a credit transfer:

Here is the error I get:

```
Payment Type Code 'pain.001.001.03.ch.02' is not supported. The only Payment Type Codes supported for SEPA Credit Transfers are 'pain.001.001.02', 'pain.001.001.03', 'pain.001.001.04', 'pain.001.001.05', 'pain.001.003.03' and 'pain.001.003.03.ch.02'. 
```

File: l10n_ch_sepa_credit_transfer/wizard/export_sepa.py [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+            name_maxsize = 140
+            root_xml_tag = 'CstmrCdtTrfInitn'
+        # added pain.001.003.03 for German Banks
+        # it is not in the offical ISO 20022 documentations, but nearly all
+        # german banks are working with this instead 001.001.03
+        elif pain_flavor == 'pain.001.003.03':
+            bic_xml_tag = 'BIC'
+            name_maxsize = 70
+            root_xml_tag = 'CstmrCdtTrfInitn'
+        elif pain_flavor == 'pain.001.003.03.ch.02':
```
Comment (yvaucher) [inline l10n_ch_sepa_credit_transfer/wizard/export_sepa.py]: I guess there is a typo here:

xml data file is defining: `pain.001.001.03.ch.02`

**Review by yvaucher — CHANGES_REQUESTED** [review]

File: l10n_ch_sepa_credit_transfer/wizard/export_sepa.py [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+            'pain.001.001.03',
+            'pain.001.001.04',
+            'pain.001.001.05',
+            'pain.001.003.03'
+        ]
+
+        # A. Group header
+        group_header_1_0, nb_of_transactions_1_6, control_sum_1_7 = \
+            self.generate_group_header_block(
+                cr, uid, pain_root, gen_args, context=context)
```
Comment (yvaucher) [inline l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:311]: Here, `gen_args` is undefined if pain_flavor == 'pain.001.003.03.ch.02'

**Review by yvaucher — CHANGES_REQUESTED** [review]

File: l10n_ch_sepa_credit_transfer/wizard/export_sepa.py (hunk 9 above)
Comment (abaghirli) [reply l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:426]: i agree. But i was afraid to remove this part of code:
https://github.com/OCA/bank-payment/blob/7.0/account_banking_sepa_credit_transfer/wizard/export_sepa.py#L278

File: l10n_ch_sepa_credit_transfer/wizard/export_sepa.py [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+                    "are 'pain.001.001.02', 'pain.001.001.03', "
+                    "'pain.001.001.04', 'pain.001.001.05',"
+                    " 'pain.001.003.03' and 'pain.001.003.03.ch.02'.") %
+                pain_flavor)
+
+        if pain_flavor == 'pain.001.003.03.ch.02':
+            module = 'l10n_ch_sepa_credit_transfer'
+        else:
+            module = 'account_banking_sepa_credit_transfer'
+            gen_args = {
```
Comment (yvaucher) [inline l10n_ch_sepa_credit_transfer/wizard/export_sepa.py]: Indentation is also wrong here

File: l10n_ch_sepa_credit_transfer/wizard/export_sepa.py [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+        '''
+        Creates the SEPA Credit Transfer file. That's the important code !
+        '''
+        if context is None:
+            context = {}
+        sepa_export = self.browse(cr, uid, ids[0], context=context)
+        pain_flavor = sepa_export.payment_order_ids[0].mode.type.code
+        convert_to_ascii = \
+            sepa_export.payment_order_ids[0].mode.convert_to_ascii
+        if pain_flavor == 'pain.001.001.02':
```
Comment (simahawk) [inline l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:232]: these are all static values: move them to a mapping. For instance:
```python
class BankingExportSepaWizard(...):
    pain_flavor_mapping = {
        'pain.001.001.02':  {...}
        'pain.001.001.03':  {...}
        [...]
    }
    [...]
    pain_values = self.pain_flavor_mapping.get(pain_flavor)
```

File: l10n_ch_sepa_credit_transfer/wizard/export_sepa.py [hunk 18]
```diff
@@ -0,0 +1,303 @@
...
+        elif pain_flavor == 'pain.001.001.03.ch.02':
+            bic_xml_tag = 'BIC'
+            name_maxsize = 70
+            root_xml_tag = 'CstmrCdtTrfInitn'
+        else:
+            raise orm.except_orm(
+                _('Error:'),
+                _("Payment Type Code '%s' is not supported. The only "
+                    "Payment Type Codes supported for SEPA Credit Transfers "
+                    "are 'pain.001.001.02', 'pain.001.001.03', "
```
Comment (simahawk) [inline l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:272]: according to the change request above:
```python
"are %s" % ', '.join(self.pain_flavor_mapping.keys())
```

File: l10n_ch_sepa_credit_transfer/wizard/export_sepa.py [hunk 19]
```diff
@@ -0,0 +1,303 @@
...
+                cr, uid, pain_root, gen_args, context=context)
+
+        transactions_count_1_6 = 0
+        total_amount = 0.0
+        amount_control_sum_1_7 = 0.0
+        lines_per_group = {}
+        # key = (requested_date, priority)
+        # values = list of lines as object
+        today = fields.date.context_today(self, cr, uid, context=context)
+        for payment_order in sepa_export.payment_order_ids:
```
Comment (simahawk) [inline l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:320]: split this to a separate method `group_lines`

File: l10n_ch_sepa_credit_transfer/wizard/export_sepa.py [hunk 20]
```diff
@@ -0,0 +1,303 @@
...
+                    requested_date = payment_order.date_scheduled or today
+                else:
+                    requested_date = today
+                key = (requested_date, priority)
+                if key in lines_per_group:
+                    lines_per_group[key].append(line)
+                else:
+                    lines_per_group[key] = [line]
+                # Write requested_date on 'Payment date' of the pay line
+                if requested_date != line.date:
```
Comment (simahawk) [inline l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:337]: if we expect many lines here is better to collect a list of `(line_id, date)` to write everything at the end w/ a single query.
See [this example](https://github.com/camptocamp/bso_odoo/blob/master/odoo/local-src/sale_purchase_sourcing/wizards/wiz_sale_order_source.py#L285).

File: l10n_ch_sepa_credit_transfer/wizard/export_sepa.py [hunk 21]
```diff
@@ -0,0 +1,303 @@
...
+                'sepa_export.payment_order_ids[0].mode.bank_id.bank.bic',
+                {'sepa_export': sepa_export},
+                gen_args, context=context)
+
+            charge_bearer_2_24 = etree.SubElement(payment_info_2_0, 'ChrgBr')
+            charge_bearer_2_24.text = sepa_export.charge_bearer
+
+            transactions_count_2_4 = 0
+            amount_control_sum_2_5 = 0.0
+            for line in lines:
```
Comment (simahawk) [inline l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:370]: split this to a method too and pass it only required args like the counters

File: l10n_ch_sepa_credit_transfer/wizard/export_sepa.py [hunk 22]
```diff
@@ -0,0 +1,303 @@
...
+                    party_account_iban, 'Id')
+                party_account_othr_id.text = viban
+        if order == 'B':
+            self.generate_party_agent(
+                cr, uid, parent_node, party_type, party_type_label,
+                order, party_name, viban, bic,
+                eval_ctx, gen_args, context=context)
+        return True
+
+    def create_sepa(self, cr, uid, ids, context=None):
```
Comment (simahawk) [inline l10n_ch_sepa_credit_transfer/wizard/export_sepa.py]: :stop_sign:  this method is way too long: more than 200 lines. Hard to read, understand and test.
See my suggestions for splitting it into helper methods.

File: l10n_ch_sepa_direct_debit/wizard/export_sdd.py [hunk 23]
```diff
@@ -0,0 +1,398 @@
...
+                    party_account_iban, 'Id')
+                party_account_othr_id.text = viban
+        if order == 'B':
+            self.generate_party_agent(
+                cr, uid, parent_node, party_type, party_type_label,
+                order, party_name, viban, bic,
+                eval_ctx, gen_args, context=context)
+        return True
+
+    def create_sepa(self, cr, uid, ids, context=None):
```
Comment (simahawk) [inline l10n_ch_sepa_direct_debit/wizard/export_sdd.py:103]: same as above: way too long. Should be organized into meaningful blocks delegated to helper methods.

**Review by simahawk — CHANGES_REQUESTED** [review]

**Comment by simahawk** [comment]
BTW, no tests?

File: l10n_ch_sepa_direct_debit/wizard/export_sdd.py (hunk 23 above)
Comment (yvaucher) [reply l10n_ch_sepa_direct_debit/wizard/export_sdd.py:103]: @simahawk this is base on a port of a module from version 8 or 9 we probably don't want to change it too much.

**Pull request closed without merging**